imaginex_lambda/handler.py

```diff
@@ -85,42 +90,61 @@ def handler(event, context):
 
         print(url, width, quality)
 
-        if not url:
-            return error('url is required')
-        if width <= 0:
-            return error('width must be greater than zero')
-
-        with TemporaryFile() as buffer:
-            if is_absolute(url):
-                download_image(buffer, url)
-            else:
-                key = url.strip('/')
-                get_s3_image(buffer, key)
-
-            original = os.stat(buffer.name).st_size
-            mime = get_extension(buffer)
-            content_type = mime['content_type']
-            extension = mime['extension']
-
-            image_data = optimize_image(
-                buffer,
-                ext=extension,
-                width=width,
-                quality=quality
-            )
-
-        print("Returning data...")
+        image_data, content_type, optimization_ratio = download_and_optimize(url, quality, width)
+
         return success(image_data, {
             'Vary': 'Accept',
             'Content-Type': content_type,
-            'X-Optimization-Ratio': f'{len(image_data) / original:.4f}',
+            'X-Optimization-Ratio': f'{optimization_ratio:.4f}',
         })
+    except HandlerError as exc:
+        return error(str(exc), code=exc.code)
     except Exception as exc:
         return error(str(exc), code=500)
 
 
+def download_and_optimize(url: str, quality: int, width: int):
+    """
+    This is the function responsible for coordinating the download and optimization of the images. It should
+    not concern itself with any lambda-specific information.
+    """
+
+    if not url:
+        raise HandlerError('url is required')
+    if width <= 0:
+        raise HandlerError('width must be greater than zero')
+
+    with TemporaryFile() as buffer:
+        if is_absolute(url):
+            download_image(buffer, url)
+        else:
+            key = url.strip('/')
+            get_s3_image(buffer, key)
+
+        buffer.flush()
+        original = os.stat(buffer.name).st_size
+        mime = get_extension(buffer)
+        content_type = mime['content_type']
+        extension = mime['extension']
+
+        image_data = optimize_image(
+            buffer,
+            ext=extension,
+            width=width,
+            quality=quality
+        )
+
+    ratio = len(image_data) / original if original != 0 else 0
+
+    print("Returning data...")
+    return image_data, content_type, ratio
+
+
 if __name__ == '__main__':
     print("Running test...")
 
-    data = json.load(open('example/event-absolute-jfif.json'))
-    handler(data, None)
+    context = {'queryStringParameters':
+               {'q': '40', 'w': '250',
+                'url': 'https://s3.eu-central-1.amazonaws.com/fllite-dev-main/'
```

Comment:
Let's change this one to `http://sladg-imaginex.s3.eu-central-1.amazonaws.com`
I will upload some set of example images there

Reply:
Aka. `S3_BUCKET_NAME=sladg-imaginex`

Reply:
I'd rather send a new PR with the new bucket, is that OK?

Here's a preview: https://github.com/fabiob/imaginex-lambda/tree/more-tests

Reply:
Sounds good!